src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -0,0 +1,231 @@
+              fatal(
+                s"""invalid sample ids: expected sample ids to be identical for all inputs. Found different sample ids at position $j.
+                   |    ${files(0)}: $s1
+                   |    ${fileByPartition(i)}: $s2""".
+                  stripMargin)
+            }
+          }
+        }
+      }
+
+      val region = MemoryBuffer()
+      val rvb = new RegionValueBuilder(region)
+      val rv = RegionValue(region)
+
+      new Iterator[RegionValue] {
+        var present = false
+
+        def advance() {
+          while (!present && it.hasNext) {
+            it.next().foreach { line =>
+              if (line.nonEmpty) {
+                val k = line.substring(0,line.indexOf(sep))
+                region.clear()
+                rvb.start(keyType)
+                rvb.startStruct()
+                rvb.addString(k)
+                rvb.addString(k)
+                rvb.endStruct()
+                rv.setOffset(rvb.end())
+
+                present = true
+              }
+            }
+          }
+        }
+
+        def hasNext: Boolean = {
+          if (!present)
+            advance()
+          present
+        }
+
+        def next(): RegionValue = {
+          hasNext
+          assert(present)
+          present = false
+          rv
+        }
+      }
+    }
+
+    val rdd = lines
+      .mapPartitionsWithIndex { (i,it) =>
+
+        val region = MemoryBuffer()
+        val rvb = new RegionValueBuilder(region)
+        val rv = RegionValue(region, 0)
+
+        if (firstPartitions.contains(i)) {it.next()}
+
+        new Iterator[RegionValue] {
+          var present = false
+
+          def advance() {
+            while (!present && it.hasNext) {
+              it.next().foreach { line =>
+                if (line.nonEmpty) {
+
+                  val firstsep = line.indexOf(sep)
+
+                  region.clear()
+                  rvb.start(matrixType.orderedRDD2Type.rowType.fundamentalType)
+                  rvb.startStruct()
+
+                  rvb.addString(line.substring(0,firstsep))
+                  rvb.addString(line.substring(0,firstsep))
+                  rvb.startStruct()
+                  rvb.endStruct()
+
+                  rvb.startArray(nSamples)
+                  if (nSamples > 0) {
+                    var off = firstsep + 1
+                    var v = 0L
+                    var ii = 0
+                    while (ii < nSamples) {
+                      if (off > line.length) {
+                        fatal(
+                          s"""Incorrect number of elements in line:
+                             |    expected $nSamples elements in row but only $i elements found.
+                             |    in file ${fileByPartition(i)}
+                           """.stripMargin
+                        )
+                      }
+                      if (off == line.length || line(off) == sep(0)) {
+                        rvb.addLong(v)
```

Comment:
This treats empty elements as 0?

Reply:
Should I be using setMissing() to designate empty elements? How about if a line doesn't have the right number of /t separators? (Currently it throws an error)

Reply:
empty elements should be an error (unless the missing string is itself empty, although it normally defaults to NA).  Missing elements should use setMissing, yes.  I think throwing an error for the wrong  number of columns is good.

Reply:
How do I differentiate between a missing element and an empty element? Like if I'm parsing a line that goes ... 0\t\t1\t0\t ... that should throw an error? But what does a missing element look like?

Reply:
There should be an `missing='NA'` option (like `HailContext.import_table`, see my other comment) to `import_matrix` that says what the missing element text should be.  (`missing=None` should mean no missingness.)

So if `missing='NA'`, then yes, `...0\t\t1\t0\t...` should throw an error.  And `...0\tNA\t1\t0\t...` should be OK (with a missing element).

Reply:
(never mind, I just saw your comment below. I'll take a look at the import_table stuff on missingness first.)